[SCEV] Support addrec in right hand side in howManyLessThans

mrdaybird opened the pull request:
Fixes #92554 (std::reverse will auto-vectorize now)

When calculating number of times a exit condition containing a comparison is executed, we mostly assume that RHS of comparison should be loop invariant, but it may be another add-recurrence.

~In that case, we can try the computation with `LHS = LHS - RHS` and `RHS = 0`.~ (It is not valid unless proven that it doesn't wrap) 

**Edit:**
We can calculate back edge count for loop structure like:

```cpp
left = left_start
right = right_start
while(left < right){
  // ...do something...
  left += s1; // the stride of left is s1 (> 0)
  right -= s2; // the stride of right is -s2 (s2 > 0)
}
// left and right converge somewhere in the middle of their start values
```
We can calculate the backedge-count as ceil((End - left_start) /u (s1- (-s2)) where, End = max(left_start, right_start).

**Alive2**: https://alive2.llvm.org/ce/z/ggxx58

mrdaybird committed:
Update howManyLessThans

mrdaybird commented:
I had a few queries:
1. What is the need of `howManyGreaterThans`, can we not reuse `howManyLessThans` to do the same computations?
2. The comment in [line 9198](https://github.com/llvm/llvm-project/blob/9b6f0735fbff4e9a51f0c76f5630be649490f6ef/llvm/lib/Analysis/ScalarEvolution.cpp#L9198) mentions that RHS cannot include boundary value. Is this still true if the RHS is an add-recurrence? For example, consider that LHS is a increasing recurrence and RHS is decreasing recurrence.

@nikic @fhahn

efriedma-quic commented:
> What is the need of howManyGreaterThans, can we not reuse howManyLessThans to do the same computations?

They probably should be merged... but it's not trivial.  If you just blindly NOT both sides, the result would be correct, but probably less accurate.  Otherwise, you need to add a bunch of conditionals to track whether you're supposed to be incrementing or decrementing.

> The comment in [line 9198](https://github.com/llvm/llvm-project/blob/9b6f0735fbff4e9a51f0c76f5630be649490f6ef/llvm/lib/Analysis/ScalarEvolution.cpp#L9198) mentions that RHS cannot include boundary value. Is this still true if the RHS is an add-recurrence?

In general, if the RHS is loop-variant, the LHS can potentially wrap even if the loop isn't infinite.  I don't think restricting to add-recurrences on the RHS changes that.

efriedma-quic commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -12941,12 +12941,16 @@ ScalarEvolution::howManyLessThans(const SCEV *LHS, const SCEV *RHS,
-  // bound of the loop (RHS), and the fact that IV does not overflow (which is
-  // checked above).
   if (!isLoopInvariant(RHS, L)) {
+    // If RHS is an add recurrence, try again with lhs=lhs-rhs and rhs=0
```
You can't subtract a value from the LHS and RHS of the comparison unless you can prove the subtraction doesn't wrap.

mrdaybird replied:
@efriedma-quic Oh!
So suppose, we can prove that LHS Start < RHS Start and we know stride of lhs is positive and stride of RHS is negative then it should do the trick right?

efriedma-quic commented:
I don't see anything obviously wrong in your example; probably something simple is going wrong.  Are you sure the expressions you're computing match the condition? Maybe try setting a breakpoint on ScalarEvolution::isImpliedCondBalancedTypes and step through.

mrdaybird committed:
Update howManyLessThans

mrdaybird committed:
Add negative rhs stride condition

mrdaybird commented:
@efriedma-quic I have modified the code, maybe you can have a look at it?
If everything's alright, then I will go ahead and add some tests.

efriedma-quic commented:
The general approach here seems fine; there are some special-cases you could optimize, but we can leave that for a followup.

Please refactor the copy-pasted code so the relevant codepaths are shared.

mrdaybird committed:
Fix formatting

mrdaybird commented:
@efriedma-quic I copy-pasted the return logic because otherwise I had to push everything into a else branch(which would be pain to review and git-diff :sweat_smile:)

efriedma-quic commented:
Maybe you can flatten the control flow using the `if (!BECount) {` checks?

Otherwise, it's not a big deal; we can always review the diff with whitespace turned off.

hiraditya commented:
Thanks for putting this patch! You might want to use https://alive2.llvm.org/ce/ to verify transformations. It is quite useful in finding corner cases.

nikic commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -12941,179 +12941,223 @@ ScalarEvolution::howManyLessThans(const SCEV *LHS, const SCEV *RHS,
-      if (isLoopEntryGuardedByCond(L, CondGE, OrigRHS, OrigStart) ||
-          isKnownPredicate(CondGE, GuardedRHS, GuardedStart))
-        return true;
+    if (auto RHSAddRec = dyn_cast<SCEVAddRecExpr>(RHS)) {
```
```suggestion
    if (const auto *RHSAddRec = dyn_cast<SCEVAddRecExpr>(RHS)) {
```

nikic commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -12941,179 +12941,223 @@ ScalarEvolution::howManyLessThans(const SCEV *LHS, const SCEV *RHS,
+      // check if RHSStride<0 and Stride-RHSStride will not overflow
+      // FIXME: Can RHSStride be positive?
+      if (isKnownNegative(RHSStride) &&
+          willNotOverflow(llvm::Instruction::Sub, true, Stride, RHSStride)) {
```
```suggestion
          willNotOverflow(Instruction::Sub, /*Signed=*/true, Stride, RHSStride)) {
```

nikic reviewed (commented):
Can you please add some tests in llvm/test/Analysis/ScalarEvolution?

mrdaybird committed:
Add tests

mrdaybird commented:
Alive2 proof: https://alive2.llvm.org/ce/z/pFJeUx -> checks for validity for all start values as well as all strides(i.e. s1, s2 such that s1>0 and s2<0 and (s1-s2) does not overflow)

mrdaybird commented:
@nikic I have added tests and verified the code using alive2. I hope everything alright!

efriedma-quic commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -12941,179 +12941,223 @@ ScalarEvolution::howManyLessThans(const SCEV *LHS, const SCEV *RHS,
-      if (isLoopEntryGuardedByCond(L, CondGE, OrigRHS, OrigStart) ||
-          isKnownPredicate(CondGE, GuardedRHS, GuardedStart))
-        return true;
+    if (const auto *RHSAddRec = dyn_cast<SCEVAddRecExpr>(RHS)) {
```
Do you need to check `RHS->getLoop() == L`?

mrdaybird replied:
I missed this. Though it may be implied given the context, but I couldn't prove(or disprove) it. Nonetheless, I have added the condition in the revised code.

mrdaybird committed:
Add condition for RHS to be in the same loop, fix formatting

mrdaybird commented:
@efriedma-quic I hope the PR looks alright. If there is any room for improvement, please do tell!

efriedma-quic commented:
Please revise the Alive2 proof so it doesn't assume the loop is guarded by `icmp slt i8 %left_start, %right_start`; the code in the patch doesn't check for that.  (I think your patch works correctly in that case, but it would be nice to have the proof.)

-----

Please take a bit of time to clean up the comments (capitalize the first word of a sentence, use "//" for block comments).

mrdaybird committed:
Update comment formatting

mrdaybird commented:
@efriedma-quic Updated alive2 proof: https://alive2.llvm.org/ce/z/sXwLZY. 
Also, fixed the comments.

hiraditya commented:
> Then number of iterations will be (right_start - left_start)/(s1-(-s2)).

What is the number of iterations when s1+s2 can't exactly divide (right_start - left_start)? It might be a `(right_start - left_start)/(s1-(-s2)) + 1`

mrdaybird commented:
> > Then number of iterations will be (right_start - left_start)/(s1-(-s2)).
> 
> What is the number of iterations when s1+s2 can't exactly divide (right_start - left_start)? It might be a `(right_start - left_start)/(s1-(-s2)) + 1`

In the code, backedge count is calculated as ceil((End - Start) /u (Stride - RHSStride) where, End = max(RHSStart, Start).

efriedma-quic commented:
I was experimenting with the Alive2 proof a bit, and ran into the following which seems a bit worrying: https://alive2.llvm.org/ce/z/gabymq .  (See https://github.com/AliveToolkit/alive2/issues/727 for discussion of the unrolling options.)

hiraditya commented:
> > > Then number of iterations will be (right_start - left_start)/(s1-(-s2)).
> > 
> > 
> > What is the number of iterations when s1+s2 can't exactly divide (right_start - left_start)? It might be a `(right_start - left_start)/(s1-(-s2)) + 1`
> 
> In the code, backedge count is calculated as ceil((End - left_start) /u (s1- (-s2)) where, End = max(left_start, right_start).

Yeah i figured, but the comment was misleading. Thanks for fixing the commit message.

mrdaybird commented:
> I was experimenting with the Alive2 proof a bit, and ran into the following which seems a bit worrying: https://alive2.llvm.org/ce/z/gabymq . (See [AliveToolkit/alive2#727](https://github.com/AliveToolkit/alive2/issues/727) for discussion of the unrolling options.)

Oh! This indeed is a bug that needs to be handled. 
I am thinking that, adding a checking for any no-wrap flags for RHSAddRec should suffice at minimum, though there could be more cases that we can handle. What do you think? @efriedma-quic 
Also alive2 also seems to not have any problems if the RHS recurrence has nsw/nuw flags.

efriedma-quic commented:
There are probably some other cases we could handle... but just checking the RHS doesn't wrap seems fine for now.

mrdaybird committed:
Check for RHSAddRec nowrap flags and add test

mrdaybird commented:
@efriedma-quic Added the rhs wrap check. Alive2 proof:https://alive2.llvm.org/ce/z/f2CpXv
For future reference, what other cases do you think we could handle here?

hiraditya commented:
We should have a test where s1 != s2 e.g., s1=3, s2=4.

efriedma-quic commented:
> For future reference, what other cases do you think we could handle here?

I think you can special-case the case where the positive and negative strides are equal: if the arithmetic doesn't overflow immediately, it can't overflow at all.  Maybe you can generalize that a bit...

mrdaybird commented:
@efriedma-quic Currently we are checking if the RHSStride<0, but alive2 shows that only requirement is that the strides should not be equal to each other. (https://alive2.llvm.org/ce/z/vPvwri). Should I trust it and incorporate the same in the code?
(Also, I know that tests are failing; due a stupid mistake by me.)

mrdaybird committed:
Add test and fix last commit

mrdaybird commented:
Anyway, I have fixed the test error. Also, I added the check that Stride(LHS) should be Positive, I mistakenly assumed it was implied. 
For the current code, (left_stride > 0 and right_stride < 0), Alive2 proof(with unrolling): https://alive2.llvm.org/ce/z/f2CpXv

efriedma-quic commented:
If the strides are equal, you end up dividing by zero.  But in that case, the actual backedge-taken count must be 0 (because otherwise it would wrap).  So you can actually fix that case by just using `umax(s1-s2, 1)` as the denominator

Whether it's worth supporting arbitrary strides... I guess we might as well?  I mean, if we're only support nowrap cases anyway, it's not really that much harder to reason about.  We'd probably need to add back the checks if we ever try to support no-nowrap cases, though.

mrdaybird commented:
I guess I will broaden the scope and add optimisations in a follow up PR.

efriedma-quic commented:
Thinking about it a bit more, there's one thing the alive2 proof currently isn't testing: other exits from the loop (or infinite loops nested inside the loop).  In those cases, even if we prove it's impossible to exit via this exit, we need to ensure that the returned backedge-taken count is greater than the number of times the loop actually iterates.

mrdaybird commented:
@efriedma-quic I am not sure what exactly needs to be done. The case that you mentioned is true in general(i.e. for any type of loop), therefore shouldn't it be already handled by the previous code. Maybe I am misunderstanding you point here? If so, maybe if you could refer me to the original alive2 proof for computing backedge count (or something similar) it would point me in the right direction!

mrdaybird commented:
@efriedma-quic From what I understand, if we could not find the exact backedge-count(like for the cases you mentioned) we return SCEVCouldNotCompute, which is considered greater than any computable value.

efriedma-quic commented:
The case I was thinking of is something like https://alive2.llvm.org/ce/z/kzxOYg .  Alive2 thinks it's fine as long as we keep the strides-not-equal check.

And yes, we can always return SCEVCouldNotCompute() if we can't prove there's a safe value to return.

mrdaybird commented:
@efriedma-quic  Interesting case that you mentioned! (good to know that it was not a problem)

efriedma-quic reviewed (approved):
LGTM.  Please don't forget to revise the commit message before you merge.

mrdaybird commented:
@nikic I have improved the commit message. (You can update it if you want!) I hope everything looks good!

hiraditya reviewed (approved):
Let's wait until Monday in case @nikic has any comments.

mrdaybird commented:
Thanks for help (with my first PR)! @efriedma-quic and @hiraditya! Learned a lot! Hopefully I can contribute more to LLVM.
